core/src/main/java/org/apache/accumulo/core/util/threads/ThreadPools.java

```diff
@@ -40,13 +44,92 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+@SuppressFBWarnings(value = "RV_EXCEPTION_NOT_THROWN",
+    justification = "Throwing Error for it to be caught by AccumuloUncaughtExceptionHandler")
 public class ThreadPools {
 
+  public static class ExecutionError extends Error {
+
+    private static final long serialVersionUID = 1L;
+
+    public ExecutionError(String message, Throwable cause) {
+      super(message, cause);
+    }
+  }
+
   private static final Logger LOG = LoggerFactory.getLogger(ThreadPools.class);
 
   // the number of seconds before we allow a thread to terminate with non-use.
   public static final long DEFAULT_TIMEOUT_MILLISECS = 180000L;
 
+  private static final ThreadPoolExecutor SCHEDULED_FUTURE_CHECKER_POOL =
+      createFixedThreadPool(1, "Scheduled Future Checker", false);
+
+  private static final ConcurrentLinkedQueue<ScheduledFuture<?>> CRITICAL_RUNNING_TASKS =
+      new ConcurrentLinkedQueue<>();
+
+  private static Runnable CRITICAL_TASK_CHECKER = new Runnable() {
+    @Override
+    public void run() {
+      while (true) {
+        Iterator<ScheduledFuture<?>> futures = CRITICAL_RUNNING_TASKS.iterator();
+        while (futures.hasNext()) {
+          ScheduledFuture<?> criticalTask = futures.next();
+          // Calling get() on a ScheduledFuture will block unless that scheduled task has
+          // completed. We call isDone() here instead. If the scheduled task is done then
+          // either it was a one-shot task, cancelled or an exception was thrown.
+          if (criticalTask.isDone()) {
+            // Now call get() to see if we get an exception.
+            try {
+              criticalTask.get();
+              // If we get here, then a scheduled task exited but did not throw an error
+              // or get canceled. This was likely a one-shot scheduled task (I don't think
+              // we can tell if it's one-shot or not, I think we have to assume that it is
+              // and that a recurring task would not normally be complete).
```

Comment:
We could assume that one shot task should never be added to the list, so that even if it exits without an exception its still considered an error that will leave the process unhealthy.  Could be an indication of a bug in the sched task, so maybe throw an error or log an error.

Reply:
There are places in the code where we schedule a task to run once, so if it exits normally, I don't think it's an error. Here's an example: https://github.com/apache/accumulo/blob/main/server/tserver/src/main/java/org/apache/accumulo/tserver/AssignmentHandler.java#L223. If an error is thrown in the one-shot task, this code will handle it.

Reply:
Thinking more about it, I think what I was worried about can not happen.  A recurring scheduled task will never "exit", those will probably only report done if there is an error or its cancelled.

Reply:
That's right. In fact, `Future.get()` will *not* return (it will block) for a ScheduledFuture, which is why I check to see if `Future.isDone()` returns `true` first. There is this note in https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/Future.html:

```
The result can only be retrieved using method get when the computation has completed, blocking if necessary until it is ready
```